File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        Queue<Pair> queue = new LinkedList<>();
        queue.offer(new Pair(root, 0));

        int maxWidth = 1;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            int left = 0, right = 0;

            for (int i = 0; i < levelSize; i++) {
                Pair pair = queue.poll();
                TreeNode node = pair.node;
                int index = pair.index;

                if (i == 0) left = index;
                if (i == levelSize - 1) right = index;
                
                if (node.left != null) {
                    queue.offer(new Pair(node.left, 2 * index));
                }
                if (node.right != null) {
                    queue.offer(new Pair(node.right, 2 * index + 1));
                }
            }

            maxWidth = Math.max(maxWidth, right - left + 1);
        }

        return maxWidth;
    }

    class Pair {
        TreeNode node;
        int index;

        Pair(TreeNode node, int index) {
            this.node = node;
            this.index = index;
        }
    }
}

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun widthOfBinaryTree(root: TreeNode?): Int {
        if (root == null) return 0

        val queue = LinkedList<Pair<TreeNode, Int>>()
        queue.offer(root to 0)

        var maxWidth = 1
        while (queue.isNotEmpty()) {
            val levelSize = queue.size
            var left = 0
            var right = 0

            for (i in 0 until levelSize) {
                val (node, index) = queue.poll()

                if (i == 0) left = index
                if (i == levelSize - 1) right = index

                node.left?.let {
                    queue.offer(it to 2 * index)
                }
                node.right?.let {
                    queue.offer(it to 2 * index + 1)
                }
            }

            maxWidth = maxOf(maxWidth, right - left + 1)
        }

        return maxWidth
    }
}